// @shineiichijo/marika/dist/lib/jikan/characters.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Characters_fetch;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Characters = void 0;
const utils_1 = require("../../utils");
class Characters {
    /**
     * Constructs an instance of the [characters](https://docs.api.jikan.moe/#tag/characters) client
     * @param cacheOptions [Cache options](https://axios-cache-interceptor.js.org/config) for the client to make requests
     */
    constructor(cacheOptions) {
        _Characters_fetch.set(this, void 0);
        /**
         * Gets the data of a character from its MyAnimeList ID
         * @param id MyAnimeList ID of the character
         * @returns The data of the character
         */
        this.getCharacterById = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Characters_fetch, "f").call(this, (0, utils_1.getURL)('characters', `${id}`))).data;
        });
        /**
         * Gets the full data of a character from its MyAnimeList ID
         * @param id MyAnimeList ID of the character
         * @returns The full data of the character
         */
        this.getCharacterFullById = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Characters_fetch, "f").call(this, (0, utils_1.getURL)('characters', `${id}`, 'full'))).data;
        });
        /**
         * Gets the list of anime appeared by a character from its MyAnimeList ID
         * @param id MyAnimeList ID of the character
         * @returns List of anime appeared by the character
         */
        this.getCharacterAnime = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Characters_fetch, "f").call(this, (0, utils_1.getURL)('characters', `${id}`, 'anime'))).data;
        });
        /**
         * Gets the list of manga appeared by a character from its MyAnimeList ID
         * @param id MyAnimeList ID of the character
         * @returns List of manga appeared by the character
         */
        this.getCharacterManga = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Characters_fetch, "f").call(this, (0, utils_1.getURL)('characters', `${id}`, 'manga'))).data;
        });
        /**
         * Gets the voice actors of a character from its MyAnimeList ID
         * @param id MyAnimeList ID of the character
         * @returns Voice actors of the character
         */
        this.getCharacterVoiceActors = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Characters_fetch, "f").call(this, (0, utils_1.getURL)('characters', `${id}`, 'voices'))).data;
        });
        /**
         * Gets the pictures of a character from its MyAnimeList ID
         * @param id MyAnimeList ID of the character
         * @returns Pictures of the character
         */
        this.getCharacterPictures = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Characters_fetch, "f").call(this, (0, utils_1.getURL)('characters', `${id}`, 'pictures'))).data;
        });
        /**
         * Searches for a character in MyAnimeList
         * @param config Config to make the search
         * @returns The results of the search
         */
        this.getCharactersSearch = (config) => __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _Characters_fetch, "f").call(this, (0, utils_1.getURL)('characters').concat((0, utils_1.getQueryString)(config || {})));
        });
        __classPrivateFieldSet(this, _Characters_fetch, new utils_1.Fetch(cacheOptions).get, "f");
    }
}
exports.Characters = Characters;
_Characters_fetch = new WeakMap();
