// @shineiichijo/marika/dist/lib/jikan/manga.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Manga_fetch;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Manga = void 0;
const utils_1 = require("../../utils");
class Manga {
    /**
     * Constructs an instance of the [manga](https://docs.api.jikan.moe/#tag/manga) client
     * @param cacheOptions [Cache options](https://axios-cache-interceptor.js.org/config) for the client to make requests
     */
    constructor(cacheOptions) {
        _Manga_fetch.set(this, void 0);
        /**
         * Gets the data of a manga from its MyAnimeList ID
         * @param id MyAnimeList ID of the manga
         * @returns The data of the manga
         */
        this.getMangaById = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Manga_fetch, "f").call(this, (0, utils_1.getURL)('manga', `${id}`))).data;
        });
        /**
         * Gets the full data of a manga from its MyAnimeList ID
         * @param id MyAnimeList ID of the manga
         * @returns The full data of the manga
         */
        this.getMangaFullById = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Manga_fetch, "f").call(this, (0, utils_1.getURL)('manga', `${id}`, 'full'))).data;
        });
        /**
         * Gets the list of characters of a manga from its MyAnimeList ID
         * @param id MyAnimeList ID of the manga
         * @returns List of characters of the manga
         */
        this.getMangaCharacters = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Manga_fetch, "f").call(this, (0, utils_1.getURL)('manga', `${id}`, 'characters'))).data;
        });
        /**
         * Gets the list of news related to a manga from its MyAnimeList ID
         * @param id MyAnimeList ID of the manga
         * @param config Config to make the request
         * @returns List of news related to the manga
         */
        this.getMangaNews = (id, config) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return yield __classPrivateFieldGet(this, _Manga_fetch, "f").call(this, (0, utils_1.getURL)('manga', `${id}`, 'news').concat((0, utils_1.getQueryString)(config || {})));
        });
        /**
         * Gets the topics related to a manga from its MyAnimeList ID
         * @param id MyAnimeList ID of the manga
         * @returns List of topics related to the manga
         */
        this.getMangaTopics = (id, config) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Manga_fetch, "f").call(this, (0, utils_1.getURL)('manga', `${id}`, 'forum').concat((0, utils_1.getQueryString)(config || {})))).data;
        });
        /**
         * Gets the pictures of a manga from its MyAnimeList ID
         * @param id MyAnimeList ID of the manga
         * @returns Pictures of the manga
         */
        this.getMangaPictures = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Manga_fetch, "f").call(this, (0, utils_1.getURL)('manga', `${id}`, 'pictures'))).data;
        });
        /**
         * Gets the statistics of a manga from its MyAnimeList ID
         * @param id MyAnimeList ID of the manga
         * @returns Statistics of the manga
         */
        this.getMangaStatistics = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Manga_fetch, "f").call(this, (0, utils_1.getURL)('manga', `${id}`, 'statistics'))).data;
        });
        /**
         * Gets more info of a manga from its MyAnimeList ID
         * @param id MyAnimeList ID of the manga
         * @returns More info of the manga
         */
        this.getMangaMoreInfo = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Manga_fetch, "f").call(this, (0, utils_1.getURL)('manga', `${id}`, 'moreinfo'))).data;
        });
        /**
         * Gets recommendations of a manga from its MyAnimeList ID
         * @param id MyAnimeList ID of the manga
         * @returns List of recommendations from the manga
         */
        this.getMangaRecommendations = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Manga_fetch, "f").call(this, (0, utils_1.getURL)('manga', `${id}`, 'recommendations'))).data;
        });
        /**
         * Gets user updates of a manga from its MyAnimeList ID
         * @param id MyAnimeList ID of the manga
         * @param config Config to make the request
         * @returns List of user updates of the manga
         */
        this.getMangaUserUpdates = (id, config) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return yield __classPrivateFieldGet(this, _Manga_fetch, "f").call(this, (0, utils_1.getURL)('manga', `${id}`, 'userupdates').concat((0, utils_1.getQueryString)(config || {})));
        });
        /**
         * Gets the reviews of a manga from its MyAnimeList ID
         * @param id MyAnimeList ID of the manga
         * @param config Config to make the request
         * @returns
         */
        this.getMangaReviews = (id, config) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return yield __classPrivateFieldGet(this, _Manga_fetch, "f").call(this, (0, utils_1.getURL)('manga', `${id}`, 'reviews').concat((0, utils_1.getQueryString)(config || {})));
        });
        /**
         * Gets relation of a manga from its MyAnimeList ID
         * @param id MyAnimeList ID of the manga
         * @returns Relations of the manga
         */
        this.getMangaRelations = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Manga_fetch, "f").call(this, (0, utils_1.getURL)('manga', `${id}`, 'relations'))).data;
        });
        /**
         * Get external sites of a manga from its MyAnimeList ID
         * @param id MyAnimeList ID of the manga
         * @returns External sites of the manga
         */
        this.getMangaExternal = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _Manga_fetch, "f").call(this, (0, utils_1.getURL)('manga', `${id}`, 'external'))).data;
        });
        /**
         * Searches for manga in MyAnimeList
         * @param config Config to make the search
         * @returns Search results of the manga
         */
        this.getMangaSearch = (config) => __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _Manga_fetch, "f").call(this, (0, utils_1.getURL)('manga').concat((0, utils_1.getQueryString)(config || {})));
        });
        __classPrivateFieldSet(this, _Manga_fetch, new utils_1.Fetch(cacheOptions).get, "f");
    }
}
exports.Manga = Manga;
_Manga_fetch = new WeakMap();
