// @shineiichijo/marika/dist/lib/jikan/people.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _People_fetch;
Object.defineProperty(exports, "__esModule", { value: true });
exports.People = void 0;
const utils_1 = require("../../utils");
class People {
    /**
     * Constructs an instance of the [people](https://docs.api.jikan.moe/#tag/people) client
     * @param cacheOptions [Cache options](https://axios-cache-interceptor.js.org/config) for the client to make requests
     */
    constructor(cacheOptions) {
        _People_fetch.set(this, void 0);
        /**
         * Gets the data of a person from its MyAnimeList ID
         * @param id MyAnimeList ID of the person
         * @returns Data of the person
         */
        this.getPersonById = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _People_fetch, "f").call(this, (0, utils_1.getURL)('people', `${id}`))).data;
        });
        /**
         * Gets the full data of a person from its MyAnimeList ID
         * @param id MyAnimeList ID of the person
         * @returns Full data of the person
         */
        this.getPersonByIdFull = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _People_fetch, "f").call(this, (0, utils_1.getURL)('people', `${id}`, 'full'))).data;
        });
        /**
         * Gets list of the anime contributed by a person from its MyAnimeList ID
         * @param id MyAnimeList ID of the person
         * @returns List of anime contributed by the person
         */
        this.getPersonAnime = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _People_fetch, "f").call(this, (0, utils_1.getURL)('people', `${id}`, 'anime'))).data;
        });
        /**
         * Gets list of the voices featured by a person from its MyAnimeList ID
         * @param id MyAnimeList ID of the person
         * @returns List of voices featured by the person
         */
        this.getPersonVoices = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _People_fetch, "f").call(this, (0, utils_1.getURL)('people', `${id}`, 'voices'))).data;
        });
        /**
         * Gets list of the manga contributed by a person from its MyAnimeList ID
         * @param id MyAnimeList ID of the person
         * @returns List of manga contributed by the person
         */
        this.getPersonManga = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _People_fetch, "f").call(this, (0, utils_1.getURL)('people', `${id}`, 'manga'))).data;
        });
        /**
         * Gets the pictures of a person from its MyAnimeList ID
         * @param id MyAnimeList ID of the person
         * @returns Pictures of the person
         */
        this.getPersonPictures = (id) => __awaiter(this, void 0, void 0, function* () {
            if (typeof id !== 'string' && typeof id !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('id', 'string or number', typeof id));
            return (yield __classPrivateFieldGet(this, _People_fetch, "f").call(this, (0, utils_1.getURL)('people', `${id}`, 'pictures'))).data;
        });
        /**
         * Searches for person in MyAnimeList
         * @param config Config to make the search
         * @returns Results of the search
         */
        this.getPeopleSearch = (config) => __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _People_fetch, "f").call(this, (0, utils_1.getURL)('people').concat((0, utils_1.getQueryString)(config || {})));
        });
        __classPrivateFieldSet(this, _People_fetch, new utils_1.Fetch(cacheOptions).get, "f");
    }
}
exports.People = People;
_People_fetch = new WeakMap();
