// @shineiichijo/marika/dist/lib/jikan/seasons.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Seasons_fetch;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Seasons = void 0;
const utils_1 = require("../../utils");
class Seasons {
    /**
     * Constructs an instance of the [seasons](https://docs.api.jikan.moe/#tag/seasons) client
     * @param cacheOptions [Cache options](https://axios-cache-interceptor.js.org/config) for the client to make requests
     */
    constructor(cacheOptions) {
        _Seasons_fetch.set(this, void 0);
        /**
         * Gets the list of current season's anime
         * @param config Config to make the request
         * @returns List of current season's anime
         */
        this.getSeasonNow = (config) => __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _Seasons_fetch, "f").call(this, (0, utils_1.getURL)('seasons', 'now').concat((0, utils_1.getQueryString)(config || {})));
        });
        /**
         * Gets the list for given season's anime of the given year
         * @param year Year of the anime
         * @param season Season of the anime
         * @param config Config to make the request
         * @returns The list for given season's anime of the given year
         */
        this.getSeason = (year, season, config) => __awaiter(this, void 0, void 0, function* () {
            if (typeof year !== 'string' && typeof year !== 'number')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('year', 'string or number', typeof year));
            if (typeof season !== 'string')
                throw new TypeError((0, utils_1.getTypeErrorMessage)('season', 'string', typeof season));
            return yield __classPrivateFieldGet(this, _Seasons_fetch, "f").call(this, (0, utils_1.getURL)('seasons', `${year}`, `${season}`).concat((0, utils_1.getQueryString)(config || {})));
        });
        /**
         * Gets the list of available seasons of all the years
         * @returns List of available seasons of all the years
         */
        this.getSeasonsList = () => __awaiter(this, void 0, void 0, function* () { return (yield __classPrivateFieldGet(this, _Seasons_fetch, "f").call(this, (0, utils_1.getURL)('seasons'))).data; });
        /**
         * Gets the list of upcoming anime of the current season
         * @param config Config to make the request
         * @returns The list of upcoming anime of the current season
         */
        this.getSeasonUpcoming = (config) => __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _Seasons_fetch, "f").call(this, (0, utils_1.getURL)('seasons', 'upcoming').concat((0, utils_1.getQueryString)(config || {})));
        });
        __classPrivateFieldSet(this, _Seasons_fetch, new utils_1.Fetch(cacheOptions).get, "f");
    }
}
exports.Seasons = Seasons;
_Seasons_fetch = new WeakMap();
