// @shineiichijo/marika/dist/lib/jikan/top.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Top_fetch;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Top = void 0;
const utils_1 = require("../../utils");
class Top {
    /**
     * Constructs an instance of the [top](https://docs.api.jikan.moe/#tag/top) client
     * @param cacheOptions [Cache options](https://axios-cache-interceptor.js.org/config) for the client to make requests
     */
    constructor(cacheOptions) {
        _Top_fetch.set(this, void 0);
        /**
         * Gets the top list of anime in MyAnimeList
         * @param config Config to get the top list
         * @returns The top list of the anime in MyAnimeList
         */
        this.getTopAnime = (config) => __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _Top_fetch, "f").call(this, (0, utils_1.getURL)('top', 'anime').concat((0, utils_1.getQueryString)(config || {})));
        });
        /**
         * Gets the top list of manga in MyAnimeList
         * @param config Config to get the top list
         * @returns The top list of the manga in MyAnimeList
         */
        this.getTopManga = (config) => __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _Top_fetch, "f").call(this, (0, utils_1.getURL)('top', 'manga').concat((0, utils_1.getQueryString)(config || {})));
        });
        /**
         * Gets the top list of people in MyAnimeList
         * @param config Config to get the top list
         * @returns The top list of people in MyAnimeList
         */
        this.getTopPeople = (config) => __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _Top_fetch, "f").call(this, (0, utils_1.getURL)('top', 'people').concat((0, utils_1.getQueryString)(config || {})));
        });
        /**
         * Gets the top list of characters in MyAnimeList
         * @param config Config to get the top list
         * @returns The top list of characters in MyAnimeList
         */
        this.getTopCharacters = (config) => __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _Top_fetch, "f").call(this, (0, utils_1.getURL)('top', 'characters').concat((0, utils_1.getQueryString)(config || {})));
        });
        /**
         * Gets the top list of reviews in MyAnimeList
         * @param config Config to get the top list
         * @returns The top list of reviews in MyAnimeList
         */
        this.getTopReviews = (config) => __awaiter(this, void 0, void 0, function* () {
            return yield __classPrivateFieldGet(this, _Top_fetch, "f").call(this, (0, utils_1.getURL)('top', 'reviews').concat((0, utils_1.getQueryString)(config || {})));
        });
        __classPrivateFieldSet(this, _Top_fetch, new utils_1.Fetch(cacheOptions).get, "f");
    }
}
exports.Top = Top;
_Top_fetch = new WeakMap();
