// @shineiichijo/marika/dist/lib/jikan/watch.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Watch_fetch;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Watch = void 0;
const utils_1 = require("../../utils");
class Watch {
    /**
     * Constructs an instance of the [watch](https://docs.api.jikan.moe/#tag/watch) client
     * @param cacheOptions [Cache options](https://axios-cache-interceptor.js.org/config) for the client to make requests
     */
    constructor(cacheOptions) {
        _Watch_fetch.set(this, void 0);
        /**
         * Gets the list of metadata of watching recent anime episodes
         * @returns The list of metadata of watching recent anime episodes
         */
        this.getWatchRecentEpisodes = () => __awaiter(this, void 0, void 0, function* () { return (yield __classPrivateFieldGet(this, _Watch_fetch, "f").call(this, (0, utils_1.getURL)('watch', 'episodes'))).data; });
        /**
         * Gets the list of metadata of watching popular anime episodes
         * @returns The list of metadata of watching popular anime episodes
         */
        this.getWatchPopularEpisodes = () => __awaiter(this, void 0, void 0, function* () { return (yield __classPrivateFieldGet(this, _Watch_fetch, "f").call(this, (0, utils_1.getURL)('watch', 'episodes', 'popular'))).data; });
        /**
         * Gets the list of metadata of watching recent anime promos
         * @returnsThe list of metadata of watching recent anime promos
         */
        this.getWatchRecentPromos = () => __awaiter(this, void 0, void 0, function* () { return (yield __classPrivateFieldGet(this, _Watch_fetch, "f").call(this, (0, utils_1.getURL)('watch', 'promos'))).data; });
        /**
         * Gets the list of metadata of watching popular anime promos
         * @returns The list of metadata of watching popular anime promos
         */
        this.getWatchPopularPromos = () => __awaiter(this, void 0, void 0, function* () { return (yield __classPrivateFieldGet(this, _Watch_fetch, "f").call(this, (0, utils_1.getURL)('watch', 'promos', 'popular'))).data; });
        __classPrivateFieldSet(this, _Watch_fetch, new utils_1.Fetch(cacheOptions).get, "f");
    }
}
exports.Watch = Watch;
_Watch_fetch = new WeakMap();
