// deobf/dist/modifications/renaming/variable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.blockScopedTypes = exports.Variable = void 0;
class Variable {
    /**
     * Creates a new variable.
     * @param name The variable name.
     * @param type The type of the variable.
     */
    constructor(name, type) {
        this.name = name;
        this.type = type;
        this.declarations = [];
        this.references = [];
    }
    /**
     * Returns whether the variable is block scoped.
     * @returns Whether.
     */
    isBlockScoped() {
        return exports.blockScopedTypes.has(this.type);
    }
    /**
     * Renames the variable.
     * @param newName The new variable name.
     */
    rename(newName) {
        for (const declaration of this.declarations) {
            declaration.name = newName;
        }
        for (const reference of this.references) {
            reference.name = newName;
        }
    }
}
exports.Variable = Variable;
exports.blockScopedTypes = new Set(["const", "let"]);
