// deobf/dist/modifications/proxies/scope.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Scope {
    /**
     * Creates a new scope.
     * @param node The node that created the scope.
     * @param parent The parent scope (optional).
     */
    constructor(node, parent) {
        this.node = node;
        this.parent = parent;
        this.children = new Map();
        this.proxyFunctions = new Map();
        if (this.parent) {
            this.parent.children.set(this.node, this);
        }
    }
    /**
     * Searches for a proxy function by name.
     * @param name The name of the proxy function.
     */
    findProxyFunction(name) {
        if (this.proxyFunctions.has(name)) {
            return this.proxyFunctions.get(name);
        }
        return this.parent ? this.parent.findProxyFunction(name) : null;
    }
    /**
     * Adds a proxy function.
     * @param proxyFunction The proxy function to be added.
     */
    addProxyFunction(proxyFunction) {
        this.proxyFunctions.set(proxyFunction.name, proxyFunction);
    }
    /**
     * Adds an alias for a proxy function.
     * @param func The proxy function.
     * @param name The alias.
     */
    addAlias(func, name) {
        this.proxyFunctions.set(name, func);
    }
}
exports.default = Scope;
