// shift-reducer/dist/adapt.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } }; // Generated by generate-adapt.js
/**
 * Copyright 2018 Shape Security, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var _shiftAst = require('shift-ast');

var Shift = _interopRequireWildcard(_shiftAst);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

exports.default = function (fn, reducer) {
  var _obj;

  return _obj = {
    __proto__: reducer,

    reduceArrayAssignmentTarget: function reduceArrayAssignmentTarget(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceArrayAssignmentTarget', this).call(this, node, data), node);
    },
    reduceArrayBinding: function reduceArrayBinding(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceArrayBinding', this).call(this, node, data), node);
    },
    reduceArrayExpression: function reduceArrayExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceArrayExpression', this).call(this, node, data), node);
    },
    reduceArrowExpression: function reduceArrowExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceArrowExpression', this).call(this, node, data), node);
    },
    reduceAssignmentExpression: function reduceAssignmentExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceAssignmentExpression', this).call(this, node, data), node);
    },
    reduceAssignmentTargetIdentifier: function reduceAssignmentTargetIdentifier(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceAssignmentTargetIdentifier', this).call(this, node, data), node);
    },
    reduceAssignmentTargetPropertyIdentifier: function reduceAssignmentTargetPropertyIdentifier(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceAssignmentTargetPropertyIdentifier', this).call(this, node, data), node);
    },
    reduceAssignmentTargetPropertyProperty: function reduceAssignmentTargetPropertyProperty(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceAssignmentTargetPropertyProperty', this).call(this, node, data), node);
    },
    reduceAssignmentTargetWithDefault: function reduceAssignmentTargetWithDefault(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceAssignmentTargetWithDefault', this).call(this, node, data), node);
    },
    reduceAwaitExpression: function reduceAwaitExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceAwaitExpression', this).call(this, node, data), node);
    },
    reduceBinaryExpression: function reduceBinaryExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceBinaryExpression', this).call(this, node, data), node);
    },
    reduceBindingIdentifier: function reduceBindingIdentifier(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceBindingIdentifier', this).call(this, node, data), node);
    },
    reduceBindingPropertyIdentifier: function reduceBindingPropertyIdentifier(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceBindingPropertyIdentifier', this).call(this, node, data), node);
    },
    reduceBindingPropertyProperty: function reduceBindingPropertyProperty(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceBindingPropertyProperty', this).call(this, node, data), node);
    },
    reduceBindingWithDefault: function reduceBindingWithDefault(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceBindingWithDefault', this).call(this, node, data), node);
    },
    reduceBlock: function reduceBlock(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceBlock', this).call(this, node, data), node);
    },
    reduceBlockStatement: function reduceBlockStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceBlockStatement', this).call(this, node, data), node);
    },
    reduceBreakStatement: function reduceBreakStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceBreakStatement', this).call(this, node, data), node);
    },
    reduceCallExpression: function reduceCallExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceCallExpression', this).call(this, node, data), node);
    },
    reduceCatchClause: function reduceCatchClause(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceCatchClause', this).call(this, node, data), node);
    },
    reduceClassDeclaration: function reduceClassDeclaration(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceClassDeclaration', this).call(this, node, data), node);
    },
    reduceClassElement: function reduceClassElement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceClassElement', this).call(this, node, data), node);
    },
    reduceClassExpression: function reduceClassExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceClassExpression', this).call(this, node, data), node);
    },
    reduceCompoundAssignmentExpression: function reduceCompoundAssignmentExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceCompoundAssignmentExpression', this).call(this, node, data), node);
    },
    reduceComputedMemberAssignmentTarget: function reduceComputedMemberAssignmentTarget(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceComputedMemberAssignmentTarget', this).call(this, node, data), node);
    },
    reduceComputedMemberExpression: function reduceComputedMemberExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceComputedMemberExpression', this).call(this, node, data), node);
    },
    reduceComputedPropertyName: function reduceComputedPropertyName(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceComputedPropertyName', this).call(this, node, data), node);
    },
    reduceConditionalExpression: function reduceConditionalExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceConditionalExpression', this).call(this, node, data), node);
    },
    reduceContinueStatement: function reduceContinueStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceContinueStatement', this).call(this, node, data), node);
    },
    reduceDataProperty: function reduceDataProperty(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceDataProperty', this).call(this, node, data), node);
    },
    reduceDebuggerStatement: function reduceDebuggerStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceDebuggerStatement', this).call(this, node, data), node);
    },
    reduceDirective: function reduceDirective(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceDirective', this).call(this, node, data), node);
    },
    reduceDoWhileStatement: function reduceDoWhileStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceDoWhileStatement', this).call(this, node, data), node);
    },
    reduceEmptyStatement: function reduceEmptyStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceEmptyStatement', this).call(this, node, data), node);
    },
    reduceExport: function reduceExport(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceExport', this).call(this, node, data), node);
    },
    reduceExportAllFrom: function reduceExportAllFrom(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceExportAllFrom', this).call(this, node, data), node);
    },
    reduceExportDefault: function reduceExportDefault(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceExportDefault', this).call(this, node, data), node);
    },
    reduceExportFrom: function reduceExportFrom(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceExportFrom', this).call(this, node, data), node);
    },
    reduceExportFromSpecifier: function reduceExportFromSpecifier(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceExportFromSpecifier', this).call(this, node, data), node);
    },
    reduceExportLocalSpecifier: function reduceExportLocalSpecifier(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceExportLocalSpecifier', this).call(this, node, data), node);
    },
    reduceExportLocals: function reduceExportLocals(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceExportLocals', this).call(this, node, data), node);
    },
    reduceExpressionStatement: function reduceExpressionStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceExpressionStatement', this).call(this, node, data), node);
    },
    reduceForAwaitStatement: function reduceForAwaitStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceForAwaitStatement', this).call(this, node, data), node);
    },
    reduceForInStatement: function reduceForInStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceForInStatement', this).call(this, node, data), node);
    },
    reduceForOfStatement: function reduceForOfStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceForOfStatement', this).call(this, node, data), node);
    },
    reduceForStatement: function reduceForStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceForStatement', this).call(this, node, data), node);
    },
    reduceFormalParameters: function reduceFormalParameters(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceFormalParameters', this).call(this, node, data), node);
    },
    reduceFunctionBody: function reduceFunctionBody(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceFunctionBody', this).call(this, node, data), node);
    },
    reduceFunctionDeclaration: function reduceFunctionDeclaration(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceFunctionDeclaration', this).call(this, node, data), node);
    },
    reduceFunctionExpression: function reduceFunctionExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceFunctionExpression', this).call(this, node, data), node);
    },
    reduceGetter: function reduceGetter(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceGetter', this).call(this, node, data), node);
    },
    reduceIdentifierExpression: function reduceIdentifierExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceIdentifierExpression', this).call(this, node, data), node);
    },
    reduceIfStatement: function reduceIfStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceIfStatement', this).call(this, node, data), node);
    },
    reduceImport: function reduceImport(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceImport', this).call(this, node, data), node);
    },
    reduceImportNamespace: function reduceImportNamespace(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceImportNamespace', this).call(this, node, data), node);
    },
    reduceImportSpecifier: function reduceImportSpecifier(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceImportSpecifier', this).call(this, node, data), node);
    },
    reduceLabeledStatement: function reduceLabeledStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceLabeledStatement', this).call(this, node, data), node);
    },
    reduceLiteralBooleanExpression: function reduceLiteralBooleanExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceLiteralBooleanExpression', this).call(this, node, data), node);
    },
    reduceLiteralInfinityExpression: function reduceLiteralInfinityExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceLiteralInfinityExpression', this).call(this, node, data), node);
    },
    reduceLiteralNullExpression: function reduceLiteralNullExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceLiteralNullExpression', this).call(this, node, data), node);
    },
    reduceLiteralNumericExpression: function reduceLiteralNumericExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceLiteralNumericExpression', this).call(this, node, data), node);
    },
    reduceLiteralRegExpExpression: function reduceLiteralRegExpExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceLiteralRegExpExpression', this).call(this, node, data), node);
    },
    reduceLiteralStringExpression: function reduceLiteralStringExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceLiteralStringExpression', this).call(this, node, data), node);
    },
    reduceMethod: function reduceMethod(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceMethod', this).call(this, node, data), node);
    },
    reduceModule: function reduceModule(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceModule', this).call(this, node, data), node);
    },
    reduceNewExpression: function reduceNewExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceNewExpression', this).call(this, node, data), node);
    },
    reduceNewTargetExpression: function reduceNewTargetExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceNewTargetExpression', this).call(this, node, data), node);
    },
    reduceObjectAssignmentTarget: function reduceObjectAssignmentTarget(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceObjectAssignmentTarget', this).call(this, node, data), node);
    },
    reduceObjectBinding: function reduceObjectBinding(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceObjectBinding', this).call(this, node, data), node);
    },
    reduceObjectExpression: function reduceObjectExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceObjectExpression', this).call(this, node, data), node);
    },
    reduceReturnStatement: function reduceReturnStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceReturnStatement', this).call(this, node, data), node);
    },
    reduceScript: function reduceScript(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceScript', this).call(this, node, data), node);
    },
    reduceSetter: function reduceSetter(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceSetter', this).call(this, node, data), node);
    },
    reduceShorthandProperty: function reduceShorthandProperty(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceShorthandProperty', this).call(this, node, data), node);
    },
    reduceSpreadElement: function reduceSpreadElement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceSpreadElement', this).call(this, node, data), node);
    },
    reduceSpreadProperty: function reduceSpreadProperty(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceSpreadProperty', this).call(this, node, data), node);
    },
    reduceStaticMemberAssignmentTarget: function reduceStaticMemberAssignmentTarget(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceStaticMemberAssignmentTarget', this).call(this, node, data), node);
    },
    reduceStaticMemberExpression: function reduceStaticMemberExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceStaticMemberExpression', this).call(this, node, data), node);
    },
    reduceStaticPropertyName: function reduceStaticPropertyName(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceStaticPropertyName', this).call(this, node, data), node);
    },
    reduceSuper: function reduceSuper(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceSuper', this).call(this, node, data), node);
    },
    reduceSwitchCase: function reduceSwitchCase(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceSwitchCase', this).call(this, node, data), node);
    },
    reduceSwitchDefault: function reduceSwitchDefault(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceSwitchDefault', this).call(this, node, data), node);
    },
    reduceSwitchStatement: function reduceSwitchStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceSwitchStatement', this).call(this, node, data), node);
    },
    reduceSwitchStatementWithDefault: function reduceSwitchStatementWithDefault(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceSwitchStatementWithDefault', this).call(this, node, data), node);
    },
    reduceTemplateElement: function reduceTemplateElement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceTemplateElement', this).call(this, node, data), node);
    },
    reduceTemplateExpression: function reduceTemplateExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceTemplateExpression', this).call(this, node, data), node);
    },
    reduceThisExpression: function reduceThisExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceThisExpression', this).call(this, node, data), node);
    },
    reduceThrowStatement: function reduceThrowStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceThrowStatement', this).call(this, node, data), node);
    },
    reduceTryCatchStatement: function reduceTryCatchStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceTryCatchStatement', this).call(this, node, data), node);
    },
    reduceTryFinallyStatement: function reduceTryFinallyStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceTryFinallyStatement', this).call(this, node, data), node);
    },
    reduceUnaryExpression: function reduceUnaryExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceUnaryExpression', this).call(this, node, data), node);
    },
    reduceUpdateExpression: function reduceUpdateExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceUpdateExpression', this).call(this, node, data), node);
    },
    reduceVariableDeclaration: function reduceVariableDeclaration(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceVariableDeclaration', this).call(this, node, data), node);
    },
    reduceVariableDeclarationStatement: function reduceVariableDeclarationStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceVariableDeclarationStatement', this).call(this, node, data), node);
    },
    reduceVariableDeclarator: function reduceVariableDeclarator(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceVariableDeclarator', this).call(this, node, data), node);
    },
    reduceWhileStatement: function reduceWhileStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceWhileStatement', this).call(this, node, data), node);
    },
    reduceWithStatement: function reduceWithStatement(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceWithStatement', this).call(this, node, data), node);
    },
    reduceYieldExpression: function reduceYieldExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceYieldExpression', this).call(this, node, data), node);
    },
    reduceYieldGeneratorExpression: function reduceYieldGeneratorExpression(node, data) {
      return fn(_get(_obj.__proto__ || Object.getPrototypeOf(_obj), 'reduceYieldGeneratorExpression', this).call(this, node, data), node);
    }
  };
};
