// shift-reducer/dist/thunkify-class.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

exports.default = thunkifyClass;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

// Generated by generate-thunkify.js
/**
 * Copyright 2018 Shape Security, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

function thunkifyClass(reducerClass) {
  return function (_reducerClass) {
    _inherits(_class, _reducerClass);

    function _class() {
      _classCallCheck(this, _class);

      return _possibleConstructorReturn(this, (_class.__proto__ || Object.getPrototypeOf(_class)).apply(this, arguments));
    }

    _createClass(_class, [{
      key: "reduceArrayAssignmentTarget",
      value: function reduceArrayAssignmentTarget(node, _ref) {
        var elements = _ref.elements,
            rest = _ref.rest;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceArrayAssignmentTarget", this).call(this, node, { elements: elements.map(function (n) {
            return n == null ? null : n();
          }), rest: rest == null ? null : rest() });
      }
    }, {
      key: "reduceArrayBinding",
      value: function reduceArrayBinding(node, _ref2) {
        var elements = _ref2.elements,
            rest = _ref2.rest;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceArrayBinding", this).call(this, node, { elements: elements.map(function (n) {
            return n == null ? null : n();
          }), rest: rest == null ? null : rest() });
      }
    }, {
      key: "reduceArrayExpression",
      value: function reduceArrayExpression(node, _ref3) {
        var elements = _ref3.elements;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceArrayExpression", this).call(this, node, { elements: elements.map(function (n) {
            return n == null ? null : n();
          }) });
      }
    }, {
      key: "reduceArrowExpression",
      value: function reduceArrowExpression(node, _ref4) {
        var params = _ref4.params,
            body = _ref4.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceArrowExpression", this).call(this, node, { params: params(), body: body() });
      }
    }, {
      key: "reduceAssignmentExpression",
      value: function reduceAssignmentExpression(node, _ref5) {
        var binding = _ref5.binding,
            expression = _ref5.expression;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceAssignmentExpression", this).call(this, node, { binding: binding(), expression: expression() });
      }
    }, {
      key: "reduceAssignmentTargetIdentifier",
      value: function reduceAssignmentTargetIdentifier(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceAssignmentTargetIdentifier", this).call(this, node);
      }
    }, {
      key: "reduceAssignmentTargetPropertyIdentifier",
      value: function reduceAssignmentTargetPropertyIdentifier(node, _ref6) {
        var binding = _ref6.binding,
            init = _ref6.init;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceAssignmentTargetPropertyIdentifier", this).call(this, node, { binding: binding(), init: init == null ? null : init() });
      }
    }, {
      key: "reduceAssignmentTargetPropertyProperty",
      value: function reduceAssignmentTargetPropertyProperty(node, _ref7) {
        var name = _ref7.name,
            binding = _ref7.binding;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceAssignmentTargetPropertyProperty", this).call(this, node, { name: name(), binding: binding() });
      }
    }, {
      key: "reduceAssignmentTargetWithDefault",
      value: function reduceAssignmentTargetWithDefault(node, _ref8) {
        var binding = _ref8.binding,
            init = _ref8.init;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceAssignmentTargetWithDefault", this).call(this, node, { binding: binding(), init: init() });
      }
    }, {
      key: "reduceAwaitExpression",
      value: function reduceAwaitExpression(node, _ref9) {
        var expression = _ref9.expression;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceAwaitExpression", this).call(this, node, { expression: expression() });
      }
    }, {
      key: "reduceBinaryExpression",
      value: function reduceBinaryExpression(node, _ref10) {
        var left = _ref10.left,
            right = _ref10.right;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceBinaryExpression", this).call(this, node, { left: left(), right: right() });
      }
    }, {
      key: "reduceBindingIdentifier",
      value: function reduceBindingIdentifier(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceBindingIdentifier", this).call(this, node);
      }
    }, {
      key: "reduceBindingPropertyIdentifier",
      value: function reduceBindingPropertyIdentifier(node, _ref11) {
        var binding = _ref11.binding,
            init = _ref11.init;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceBindingPropertyIdentifier", this).call(this, node, { binding: binding(), init: init == null ? null : init() });
      }
    }, {
      key: "reduceBindingPropertyProperty",
      value: function reduceBindingPropertyProperty(node, _ref12) {
        var name = _ref12.name,
            binding = _ref12.binding;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceBindingPropertyProperty", this).call(this, node, { name: name(), binding: binding() });
      }
    }, {
      key: "reduceBindingWithDefault",
      value: function reduceBindingWithDefault(node, _ref13) {
        var binding = _ref13.binding,
            init = _ref13.init;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceBindingWithDefault", this).call(this, node, { binding: binding(), init: init() });
      }
    }, {
      key: "reduceBlock",
      value: function reduceBlock(node, _ref14) {
        var statements = _ref14.statements;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceBlock", this).call(this, node, { statements: statements.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceBlockStatement",
      value: function reduceBlockStatement(node, _ref15) {
        var block = _ref15.block;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceBlockStatement", this).call(this, node, { block: block() });
      }
    }, {
      key: "reduceBreakStatement",
      value: function reduceBreakStatement(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceBreakStatement", this).call(this, node);
      }
    }, {
      key: "reduceCallExpression",
      value: function reduceCallExpression(node, _ref16) {
        var callee = _ref16.callee,
            _arguments = _ref16.arguments;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceCallExpression", this).call(this, node, { callee: callee(), arguments: _arguments.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceCatchClause",
      value: function reduceCatchClause(node, _ref17) {
        var binding = _ref17.binding,
            body = _ref17.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceCatchClause", this).call(this, node, { binding: binding(), body: body() });
      }
    }, {
      key: "reduceClassDeclaration",
      value: function reduceClassDeclaration(node, _ref18) {
        var name = _ref18.name,
            _super = _ref18.super,
            elements = _ref18.elements;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceClassDeclaration", this).call(this, node, { name: name(), super: _super == null ? null : _super(), elements: elements.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceClassElement",
      value: function reduceClassElement(node, _ref19) {
        var method = _ref19.method;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceClassElement", this).call(this, node, { method: method() });
      }
    }, {
      key: "reduceClassExpression",
      value: function reduceClassExpression(node, _ref20) {
        var name = _ref20.name,
            _super = _ref20.super,
            elements = _ref20.elements;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceClassExpression", this).call(this, node, { name: name == null ? null : name(), super: _super == null ? null : _super(), elements: elements.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceCompoundAssignmentExpression",
      value: function reduceCompoundAssignmentExpression(node, _ref21) {
        var binding = _ref21.binding,
            expression = _ref21.expression;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceCompoundAssignmentExpression", this).call(this, node, { binding: binding(), expression: expression() });
      }
    }, {
      key: "reduceComputedMemberAssignmentTarget",
      value: function reduceComputedMemberAssignmentTarget(node, _ref22) {
        var object = _ref22.object,
            expression = _ref22.expression;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceComputedMemberAssignmentTarget", this).call(this, node, { object: object(), expression: expression() });
      }
    }, {
      key: "reduceComputedMemberExpression",
      value: function reduceComputedMemberExpression(node, _ref23) {
        var object = _ref23.object,
            expression = _ref23.expression;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceComputedMemberExpression", this).call(this, node, { object: object(), expression: expression() });
      }
    }, {
      key: "reduceComputedPropertyName",
      value: function reduceComputedPropertyName(node, _ref24) {
        var expression = _ref24.expression;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceComputedPropertyName", this).call(this, node, { expression: expression() });
      }
    }, {
      key: "reduceConditionalExpression",
      value: function reduceConditionalExpression(node, _ref25) {
        var test = _ref25.test,
            consequent = _ref25.consequent,
            alternate = _ref25.alternate;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceConditionalExpression", this).call(this, node, { test: test(), consequent: consequent(), alternate: alternate() });
      }
    }, {
      key: "reduceContinueStatement",
      value: function reduceContinueStatement(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceContinueStatement", this).call(this, node);
      }
    }, {
      key: "reduceDataProperty",
      value: function reduceDataProperty(node, _ref26) {
        var name = _ref26.name,
            expression = _ref26.expression;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceDataProperty", this).call(this, node, { name: name(), expression: expression() });
      }
    }, {
      key: "reduceDebuggerStatement",
      value: function reduceDebuggerStatement(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceDebuggerStatement", this).call(this, node);
      }
    }, {
      key: "reduceDirective",
      value: function reduceDirective(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceDirective", this).call(this, node);
      }
    }, {
      key: "reduceDoWhileStatement",
      value: function reduceDoWhileStatement(node, _ref27) {
        var body = _ref27.body,
            test = _ref27.test;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceDoWhileStatement", this).call(this, node, { body: body(), test: test() });
      }
    }, {
      key: "reduceEmptyStatement",
      value: function reduceEmptyStatement(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceEmptyStatement", this).call(this, node);
      }
    }, {
      key: "reduceExport",
      value: function reduceExport(node, _ref28) {
        var declaration = _ref28.declaration;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceExport", this).call(this, node, { declaration: declaration() });
      }
    }, {
      key: "reduceExportAllFrom",
      value: function reduceExportAllFrom(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceExportAllFrom", this).call(this, node);
      }
    }, {
      key: "reduceExportDefault",
      value: function reduceExportDefault(node, _ref29) {
        var body = _ref29.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceExportDefault", this).call(this, node, { body: body() });
      }
    }, {
      key: "reduceExportFrom",
      value: function reduceExportFrom(node, _ref30) {
        var namedExports = _ref30.namedExports;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceExportFrom", this).call(this, node, { namedExports: namedExports.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceExportFromSpecifier",
      value: function reduceExportFromSpecifier(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceExportFromSpecifier", this).call(this, node);
      }
    }, {
      key: "reduceExportLocalSpecifier",
      value: function reduceExportLocalSpecifier(node, _ref31) {
        var name = _ref31.name;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceExportLocalSpecifier", this).call(this, node, { name: name() });
      }
    }, {
      key: "reduceExportLocals",
      value: function reduceExportLocals(node, _ref32) {
        var namedExports = _ref32.namedExports;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceExportLocals", this).call(this, node, { namedExports: namedExports.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceExpressionStatement",
      value: function reduceExpressionStatement(node, _ref33) {
        var expression = _ref33.expression;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceExpressionStatement", this).call(this, node, { expression: expression() });
      }
    }, {
      key: "reduceForAwaitStatement",
      value: function reduceForAwaitStatement(node, _ref34) {
        var left = _ref34.left,
            right = _ref34.right,
            body = _ref34.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceForAwaitStatement", this).call(this, node, { left: left(), right: right(), body: body() });
      }
    }, {
      key: "reduceForInStatement",
      value: function reduceForInStatement(node, _ref35) {
        var left = _ref35.left,
            right = _ref35.right,
            body = _ref35.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceForInStatement", this).call(this, node, { left: left(), right: right(), body: body() });
      }
    }, {
      key: "reduceForOfStatement",
      value: function reduceForOfStatement(node, _ref36) {
        var left = _ref36.left,
            right = _ref36.right,
            body = _ref36.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceForOfStatement", this).call(this, node, { left: left(), right: right(), body: body() });
      }
    }, {
      key: "reduceForStatement",
      value: function reduceForStatement(node, _ref37) {
        var init = _ref37.init,
            test = _ref37.test,
            update = _ref37.update,
            body = _ref37.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceForStatement", this).call(this, node, { init: init == null ? null : init(), test: test == null ? null : test(), update: update == null ? null : update(), body: body() });
      }
    }, {
      key: "reduceFormalParameters",
      value: function reduceFormalParameters(node, _ref38) {
        var items = _ref38.items,
            rest = _ref38.rest;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceFormalParameters", this).call(this, node, { items: items.map(function (n) {
            return n();
          }), rest: rest == null ? null : rest() });
      }
    }, {
      key: "reduceFunctionBody",
      value: function reduceFunctionBody(node, _ref39) {
        var directives = _ref39.directives,
            statements = _ref39.statements;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceFunctionBody", this).call(this, node, { directives: directives.map(function (n) {
            return n();
          }), statements: statements.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceFunctionDeclaration",
      value: function reduceFunctionDeclaration(node, _ref40) {
        var name = _ref40.name,
            params = _ref40.params,
            body = _ref40.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceFunctionDeclaration", this).call(this, node, { name: name(), params: params(), body: body() });
      }
    }, {
      key: "reduceFunctionExpression",
      value: function reduceFunctionExpression(node, _ref41) {
        var name = _ref41.name,
            params = _ref41.params,
            body = _ref41.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceFunctionExpression", this).call(this, node, { name: name == null ? null : name(), params: params(), body: body() });
      }
    }, {
      key: "reduceGetter",
      value: function reduceGetter(node, _ref42) {
        var name = _ref42.name,
            body = _ref42.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceGetter", this).call(this, node, { name: name(), body: body() });
      }
    }, {
      key: "reduceIdentifierExpression",
      value: function reduceIdentifierExpression(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceIdentifierExpression", this).call(this, node);
      }
    }, {
      key: "reduceIfStatement",
      value: function reduceIfStatement(node, _ref43) {
        var test = _ref43.test,
            consequent = _ref43.consequent,
            alternate = _ref43.alternate;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceIfStatement", this).call(this, node, { test: test(), consequent: consequent(), alternate: alternate == null ? null : alternate() });
      }
    }, {
      key: "reduceImport",
      value: function reduceImport(node, _ref44) {
        var defaultBinding = _ref44.defaultBinding,
            namedImports = _ref44.namedImports;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceImport", this).call(this, node, { defaultBinding: defaultBinding == null ? null : defaultBinding(), namedImports: namedImports.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceImportNamespace",
      value: function reduceImportNamespace(node, _ref45) {
        var defaultBinding = _ref45.defaultBinding,
            namespaceBinding = _ref45.namespaceBinding;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceImportNamespace", this).call(this, node, { defaultBinding: defaultBinding == null ? null : defaultBinding(), namespaceBinding: namespaceBinding() });
      }
    }, {
      key: "reduceImportSpecifier",
      value: function reduceImportSpecifier(node, _ref46) {
        var binding = _ref46.binding;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceImportSpecifier", this).call(this, node, { binding: binding() });
      }
    }, {
      key: "reduceLabeledStatement",
      value: function reduceLabeledStatement(node, _ref47) {
        var body = _ref47.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceLabeledStatement", this).call(this, node, { body: body() });
      }
    }, {
      key: "reduceLiteralBooleanExpression",
      value: function reduceLiteralBooleanExpression(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceLiteralBooleanExpression", this).call(this, node);
      }
    }, {
      key: "reduceLiteralInfinityExpression",
      value: function reduceLiteralInfinityExpression(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceLiteralInfinityExpression", this).call(this, node);
      }
    }, {
      key: "reduceLiteralNullExpression",
      value: function reduceLiteralNullExpression(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceLiteralNullExpression", this).call(this, node);
      }
    }, {
      key: "reduceLiteralNumericExpression",
      value: function reduceLiteralNumericExpression(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceLiteralNumericExpression", this).call(this, node);
      }
    }, {
      key: "reduceLiteralRegExpExpression",
      value: function reduceLiteralRegExpExpression(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceLiteralRegExpExpression", this).call(this, node);
      }
    }, {
      key: "reduceLiteralStringExpression",
      value: function reduceLiteralStringExpression(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceLiteralStringExpression", this).call(this, node);
      }
    }, {
      key: "reduceMethod",
      value: function reduceMethod(node, _ref48) {
        var name = _ref48.name,
            params = _ref48.params,
            body = _ref48.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceMethod", this).call(this, node, { name: name(), params: params(), body: body() });
      }
    }, {
      key: "reduceModule",
      value: function reduceModule(node, _ref49) {
        var directives = _ref49.directives,
            items = _ref49.items;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceModule", this).call(this, node, { directives: directives.map(function (n) {
            return n();
          }), items: items.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceNewExpression",
      value: function reduceNewExpression(node, _ref50) {
        var callee = _ref50.callee,
            _arguments = _ref50.arguments;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceNewExpression", this).call(this, node, { callee: callee(), arguments: _arguments.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceNewTargetExpression",
      value: function reduceNewTargetExpression(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceNewTargetExpression", this).call(this, node);
      }
    }, {
      key: "reduceObjectAssignmentTarget",
      value: function reduceObjectAssignmentTarget(node, _ref51) {
        var properties = _ref51.properties,
            rest = _ref51.rest;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceObjectAssignmentTarget", this).call(this, node, { properties: properties.map(function (n) {
            return n();
          }), rest: rest == null ? null : rest() });
      }
    }, {
      key: "reduceObjectBinding",
      value: function reduceObjectBinding(node, _ref52) {
        var properties = _ref52.properties,
            rest = _ref52.rest;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceObjectBinding", this).call(this, node, { properties: properties.map(function (n) {
            return n();
          }), rest: rest == null ? null : rest() });
      }
    }, {
      key: "reduceObjectExpression",
      value: function reduceObjectExpression(node, _ref53) {
        var properties = _ref53.properties;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceObjectExpression", this).call(this, node, { properties: properties.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceReturnStatement",
      value: function reduceReturnStatement(node, _ref54) {
        var expression = _ref54.expression;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceReturnStatement", this).call(this, node, { expression: expression == null ? null : expression() });
      }
    }, {
      key: "reduceScript",
      value: function reduceScript(node, _ref55) {
        var directives = _ref55.directives,
            statements = _ref55.statements;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceScript", this).call(this, node, { directives: directives.map(function (n) {
            return n();
          }), statements: statements.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceSetter",
      value: function reduceSetter(node, _ref56) {
        var name = _ref56.name,
            param = _ref56.param,
            body = _ref56.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceSetter", this).call(this, node, { name: name(), param: param(), body: body() });
      }
    }, {
      key: "reduceShorthandProperty",
      value: function reduceShorthandProperty(node, _ref57) {
        var name = _ref57.name;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceShorthandProperty", this).call(this, node, { name: name() });
      }
    }, {
      key: "reduceSpreadElement",
      value: function reduceSpreadElement(node, _ref58) {
        var expression = _ref58.expression;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceSpreadElement", this).call(this, node, { expression: expression() });
      }
    }, {
      key: "reduceSpreadProperty",
      value: function reduceSpreadProperty(node, _ref59) {
        var expression = _ref59.expression;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceSpreadProperty", this).call(this, node, { expression: expression() });
      }
    }, {
      key: "reduceStaticMemberAssignmentTarget",
      value: function reduceStaticMemberAssignmentTarget(node, _ref60) {
        var object = _ref60.object;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceStaticMemberAssignmentTarget", this).call(this, node, { object: object() });
      }
    }, {
      key: "reduceStaticMemberExpression",
      value: function reduceStaticMemberExpression(node, _ref61) {
        var object = _ref61.object;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceStaticMemberExpression", this).call(this, node, { object: object() });
      }
    }, {
      key: "reduceStaticPropertyName",
      value: function reduceStaticPropertyName(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceStaticPropertyName", this).call(this, node);
      }
    }, {
      key: "reduceSuper",
      value: function reduceSuper(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceSuper", this).call(this, node);
      }
    }, {
      key: "reduceSwitchCase",
      value: function reduceSwitchCase(node, _ref62) {
        var test = _ref62.test,
            consequent = _ref62.consequent;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceSwitchCase", this).call(this, node, { test: test(), consequent: consequent.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceSwitchDefault",
      value: function reduceSwitchDefault(node, _ref63) {
        var consequent = _ref63.consequent;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceSwitchDefault", this).call(this, node, { consequent: consequent.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceSwitchStatement",
      value: function reduceSwitchStatement(node, _ref64) {
        var discriminant = _ref64.discriminant,
            cases = _ref64.cases;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceSwitchStatement", this).call(this, node, { discriminant: discriminant(), cases: cases.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceSwitchStatementWithDefault",
      value: function reduceSwitchStatementWithDefault(node, _ref65) {
        var discriminant = _ref65.discriminant,
            preDefaultCases = _ref65.preDefaultCases,
            defaultCase = _ref65.defaultCase,
            postDefaultCases = _ref65.postDefaultCases;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceSwitchStatementWithDefault", this).call(this, node, { discriminant: discriminant(), preDefaultCases: preDefaultCases.map(function (n) {
            return n();
          }), defaultCase: defaultCase(), postDefaultCases: postDefaultCases.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceTemplateElement",
      value: function reduceTemplateElement(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceTemplateElement", this).call(this, node);
      }
    }, {
      key: "reduceTemplateExpression",
      value: function reduceTemplateExpression(node, _ref66) {
        var tag = _ref66.tag,
            elements = _ref66.elements;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceTemplateExpression", this).call(this, node, { tag: tag == null ? null : tag(), elements: elements.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceThisExpression",
      value: function reduceThisExpression(node) {
        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceThisExpression", this).call(this, node);
      }
    }, {
      key: "reduceThrowStatement",
      value: function reduceThrowStatement(node, _ref67) {
        var expression = _ref67.expression;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceThrowStatement", this).call(this, node, { expression: expression() });
      }
    }, {
      key: "reduceTryCatchStatement",
      value: function reduceTryCatchStatement(node, _ref68) {
        var body = _ref68.body,
            catchClause = _ref68.catchClause;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceTryCatchStatement", this).call(this, node, { body: body(), catchClause: catchClause() });
      }
    }, {
      key: "reduceTryFinallyStatement",
      value: function reduceTryFinallyStatement(node, _ref69) {
        var body = _ref69.body,
            catchClause = _ref69.catchClause,
            finalizer = _ref69.finalizer;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceTryFinallyStatement", this).call(this, node, { body: body(), catchClause: catchClause == null ? null : catchClause(), finalizer: finalizer() });
      }
    }, {
      key: "reduceUnaryExpression",
      value: function reduceUnaryExpression(node, _ref70) {
        var operand = _ref70.operand;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceUnaryExpression", this).call(this, node, { operand: operand() });
      }
    }, {
      key: "reduceUpdateExpression",
      value: function reduceUpdateExpression(node, _ref71) {
        var operand = _ref71.operand;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceUpdateExpression", this).call(this, node, { operand: operand() });
      }
    }, {
      key: "reduceVariableDeclaration",
      value: function reduceVariableDeclaration(node, _ref72) {
        var declarators = _ref72.declarators;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceVariableDeclaration", this).call(this, node, { declarators: declarators.map(function (n) {
            return n();
          }) });
      }
    }, {
      key: "reduceVariableDeclarationStatement",
      value: function reduceVariableDeclarationStatement(node, _ref73) {
        var declaration = _ref73.declaration;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceVariableDeclarationStatement", this).call(this, node, { declaration: declaration() });
      }
    }, {
      key: "reduceVariableDeclarator",
      value: function reduceVariableDeclarator(node, _ref74) {
        var binding = _ref74.binding,
            init = _ref74.init;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceVariableDeclarator", this).call(this, node, { binding: binding(), init: init == null ? null : init() });
      }
    }, {
      key: "reduceWhileStatement",
      value: function reduceWhileStatement(node, _ref75) {
        var test = _ref75.test,
            body = _ref75.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceWhileStatement", this).call(this, node, { test: test(), body: body() });
      }
    }, {
      key: "reduceWithStatement",
      value: function reduceWithStatement(node, _ref76) {
        var object = _ref76.object,
            body = _ref76.body;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceWithStatement", this).call(this, node, { object: object(), body: body() });
      }
    }, {
      key: "reduceYieldExpression",
      value: function reduceYieldExpression(node, _ref77) {
        var expression = _ref77.expression;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceYieldExpression", this).call(this, node, { expression: expression == null ? null : expression() });
      }
    }, {
      key: "reduceYieldGeneratorExpression",
      value: function reduceYieldGeneratorExpression(node, _ref78) {
        var expression = _ref78.expression;

        return _get(_class.prototype.__proto__ || Object.getPrototypeOf(_class.prototype), "reduceYieldGeneratorExpression", this).call(this, node, { expression: expression() });
      }
    }]);

    return _class;
  }(reducerClass);
}
