// @shineiichijo/marika/dist/lib/Marika.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Marika = void 0;
const jikan_1 = require("./jikan");
class Marika {
    /**
     * Constructs an instance of the [main](https://docs.api.jikan.moe/) client
     * @param cacheOptions [Cache options](https://axios-cache-interceptor.js.org/config) for the client to make requests
     */
    constructor(cacheOptions) {
        this.anime = new jikan_1.Anime(cacheOptions);
        this.characters = new jikan_1.Characters(cacheOptions);
        this.clubs = new jikan_1.Clubs(cacheOptions);
        this.genres = new jikan_1.Genres(cacheOptions);
        this.magazines = new jikan_1.Magazines(cacheOptions);
        this.manga = new jikan_1.Manga(cacheOptions);
        this.people = new jikan_1.People(cacheOptions);
        this.reviews = new jikan_1.Reviews(cacheOptions);
        this.schedules = new jikan_1.Schedules(cacheOptions);
        this.producers = new jikan_1.Producers(cacheOptions);
        this.recommendations = new jikan_1.Recommendations(cacheOptions);
        this.seasons = new jikan_1.Seasons(cacheOptions);
        this.users = new jikan_1.Users(cacheOptions);
        this.watch = new jikan_1.Watch(cacheOptions);
        this.random = new jikan_1.Random(cacheOptions);
        this.top = new jikan_1.Top(cacheOptions);
    }
}
exports.Marika = Marika;
