// shift-validator/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Validator = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

exports.default = isValid;

var _shiftParser = require("shift-parser");

var _shiftReducer = require("shift-reducer");

var _shiftReducer2 = _interopRequireDefault(_shiftReducer);

var _esutils = require("esutils");

var _shiftRegexpAcceptor = require("shift-regexp-acceptor");

var _shiftRegexpAcceptor2 = _interopRequireDefault(_shiftRegexpAcceptor);

var _validationContext = require("./validation-context");

var _validationErrors = require("./validation-errors");

var _validationErrors2 = _interopRequireDefault(_validationErrors);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2016 Shape Security, Inc.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Licensed under the Apache License, Version 2.0 (the "License")
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * you may not use this file except in compliance with the License.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * You may obtain a copy of the License at
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *     http://www.apache.org/licenses/LICENSE-2.0
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Unless required by applicable law or agreed to in writing, software
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * distributed under the License is distributed on an "AS IS" BASIS,
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * See the License for the specific language governing permissions and
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * limitations under the License.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

var isIdentifierNameES6 = _esutils.keyword.isIdentifierNameES6,
    isReservedWordES6 = _esutils.keyword.isReservedWordES6;
var isIdentifierStart = _esutils.code.isIdentifierStartES6,
    isIdentifierPart = _esutils.code.isIdentifierPartES6;
function isValid(node) {
  return Validator.validate(node).length === 0;
}

function isIterationStatement(type) {
  switch (type) {
    case "DoWhileStatement":
    case "WhileStatement":
    case "ForStatement":
    case "ForInStatement":
      return true;
  }
  return false;
}

function trailingStatement(node) {
  switch (node.type) {
    case "IfStatement":
      if (node.alternate != null) {
        return node.alternate;
      }
      return node.consequent;
    case "LabeledStatement":
    case "ForStatement":
    case "ForInStatement":
    case "ForOfStatement":
    case "WhileStatement":
    case "WithStatement":
      return node.body;
  }
  return null;
}

function isProblematicIfStatement(node) {
  if (node.type !== "IfStatement") {
    return false;
  }
  if (node.alternate == null) {
    return false;
  }
  var current = node.consequent;
  do {
    if (current.type === "IfStatement" && current.alternate == null) {
      return true;
    }
    current = trailingStatement(current);
  } while (current != null);
  return false;
}

function isValidIdentifier(name) {
  return name === 'let' || name === 'yield' || name === 'await' || name === 'async' || name === 'enum' || isIdentifierNameES6(name) && !isReservedWordES6(name);
}

function isValidIdentifierName(name) {
  return name.length > 0 && isIdentifierStart(name.charCodeAt(0)) && Array.prototype.every.call(name, function (c) {
    return isIdentifierPart(c.charCodeAt(0));
  });
}

function isValidStaticPropertyName(name) {
  return isIdentifierNameES6(name);
}

function isTemplateElement(rawValue) {
  try {
    var tokenizer = new _shiftParser.Tokenizer('`' + rawValue + '`');
    tokenizer.lookahead = tokenizer.advance();
    var token = tokenizer.lex();
    if (token.type !== _shiftParser.TokenType.TEMPLATE) {
      return false;
    }
    return tokenizer.eof();
  } catch (e) {
    return false;
  }
}

function isDirective(rawValue) {
  var stringify = function stringify(c) {
    try {
      var tokenizer = new _shiftParser.Tokenizer(c + rawValue + c);
      tokenizer.lookahead = tokenizer.advance();
      var token = tokenizer.lex();
      if (token.type !== _shiftParser.TokenType.STRING) {
        return false;
      }
      return tokenizer.eof();
    } catch (e) {
      return false;
    }
  };
  return stringify('"') || stringify("'");
}

function checkIllegalBody(node, s) {
  var _ref = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {},
      _ref$allowFunctions = _ref.allowFunctions,
      allowFunctions = _ref$allowFunctions === undefined ? false : _ref$allowFunctions;

  if (node.body.type === 'FunctionDeclaration' && !allowFunctions) {
    return s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.FUNCTION_DECLARATION_AS_STATEMENT));
  } else if (node.body.type === 'ClassDeclaration' || node.body.type === 'FunctionDeclaration' && node.body.isGenerator || node.body.type === 'VariableDeclarationStatement' && (node.body.declaration.kind === 'let' || node.body.declaration.kind === 'const')) {
    return s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.PROPER_DECLARATION_AS_STATEMENT));
  }
  return s;
}

var Validator = exports.Validator = function (_MonoidalReducer) {
  _inherits(Validator, _MonoidalReducer);

  function Validator() {
    _classCallCheck(this, Validator);

    return _possibleConstructorReturn(this, (Validator.__proto__ || Object.getPrototypeOf(Validator)).call(this, _validationContext.ValidationContext));
  }

  _createClass(Validator, [{
    key: "reduceArrowExpression",
    value: function reduceArrowExpression(node, _ref2) {
      var params = _ref2.params,
          body = _ref2.body;

      body = node.isAsync ? body.clearAwaitExpressionsNotInAsyncContext() : body.enforceAwaitExpressionsNotInAsyncContext();
      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceArrowExpression", this).call(this, node, { params: params, body: body.enforceYields() });
      return s;
    }
  }, {
    key: "reduceAssignmentTargetIdentifier",
    value: function reduceAssignmentTargetIdentifier(node) {
      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceAssignmentTargetIdentifier", this).call(this, node);
      if (!isValidIdentifier(node.name)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_BINDING_IDENTIFIER_NAME));
      }
      return s;
    }
  }, {
    key: "reduceBindingIdentifier",
    value: function reduceBindingIdentifier(node) {
      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceBindingIdentifier", this).call(this, node);
      if (!isValidIdentifier(node.name)) {
        if (node.name == "*default*") {
          s = s.addBindingIdentifierCalledDefault(node);
        } else {
          s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_BINDING_IDENTIFIER_NAME));
        }
      }
      return s;
    }
  }, {
    key: "reduceBreakStatement",
    value: function reduceBreakStatement(node) {
      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceBreakStatement", this).call(this, node);
      if (node.label !== null && !isValidIdentifier(node.label)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_BREAK_STATEMENT_LABEL));
      }
      return s;
    }
  }, {
    key: "reduceContinueStatement",
    value: function reduceContinueStatement(node) {
      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceContinueStatement", this).call(this, node);
      if (node.label !== null && !isValidIdentifier(node.label)) {
        s = s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_CONTINUE_STATEMENT_LABEL));
      }
      return s;
    }
  }, {
    key: "reduceDoWhileStatement",
    value: function reduceDoWhileStatement(node, _ref3) {
      var body = _ref3.body,
          test = _ref3.test;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceDoWhileStatement", this).call(this, node, { body: body, test: test });
      s = checkIllegalBody(node, s);
      return s;
    }
  }, {
    key: "reduceDirective",
    value: function reduceDirective(node) {
      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceDirective", this).call(this, node);
      if (!isDirective(node.rawValue)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_DIRECTIVE));
      }
      return s;
    }
  }, {
    key: "reduceExportDefault",
    value: function reduceExportDefault(node, _ref4) {
      var body = _ref4.body;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceExportDefault", this).call(this, node, { body: body });
      if (node.body.type === 'FunctionDeclaration' || node.body.type == 'ClassDeclaration') {
        s = s.clearBindingIdentifiersCalledDefault();
      }
      return s;
    }
  }, {
    key: "reduceExportFromSpecifier",
    value: function reduceExportFromSpecifier(node) {
      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceExportFromSpecifier", this).call(this, node);
      if (!isValidIdentifierName(node.name)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_EXPORT_SPECIFIER_NAME));
      }
      if (node.exportedName !== null && !isValidIdentifierName(node.exportedName)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_EXPORTED_NAME));
      }
      return s;
    }
  }, {
    key: "reduceExportLocalSpecifier",
    value: function reduceExportLocalSpecifier(node, _ref5) {
      var name = _ref5.name;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceExportLocalSpecifier", this).call(this, node, { name: name });
      if (node.exportedName !== null && !isIdentifierNameES6(node.exportedName)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_EXPORTED_NAME));
      }
      return s;
    }
  }, {
    key: "reduceForInStatement",
    value: function reduceForInStatement(node, _ref6) {
      var left = _ref6.left,
          right = _ref6.right,
          body = _ref6.body;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceForInStatement", this).call(this, node, { left: left, right: right, body: body });
      if (node.left.type === 'VariableDeclaration') {
        if (node.left.declarators.length != 1) {
          s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.ONE_VARIABLE_DECLARATOR_IN_FOR_IN));
        }
        if (node.left.declarators.length > 0 && node.left.declarators[0].init !== null) {
          s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.NO_INIT_IN_VARIABLE_DECLARATOR_IN_FOR_IN));
        }
      }
      s = checkIllegalBody(node, s);
      return s;
    }
  }, {
    key: "reduceForOfStatement",
    value: function reduceForOfStatement(node, _ref7) {
      var left = _ref7.left,
          right = _ref7.right,
          body = _ref7.body;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceForOfStatement", this).call(this, node, { left: left, right: right, body: body });
      if (node.left.type === 'VariableDeclaration') {
        if (node.left.declarators.length != 1) {
          s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.ONE_VARIABLE_DECLARATOR_IN_FOR_OF));
        }
        if (node.left.declarators.length > 0 && node.left.declarators[0].init !== null) {
          s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.NO_INIT_IN_VARIABLE_DECLARATOR_IN_FOR_OF));
        }
      }
      s = checkIllegalBody(node, s);
      return s;
    }
  }, {
    key: "reduceForStatement",
    value: function reduceForStatement(node, _ref8) {
      var init = _ref8.init,
          update = _ref8.update,
          test = _ref8.test,
          body = _ref8.body;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceForStatement", this).call(this, node, { init: init, update: update, test: test, body: body });
      s = checkIllegalBody(node, s);
      return s;
    }
  }, {
    key: "reduceFormalParameters",
    value: function reduceFormalParameters(node, _ref9) {
      var items = _ref9.items,
          rest = _ref9.rest;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceFormalParameters", this).call(this, node, { items: items, rest: rest });
      s = s.enforceYields().enforceAwaitExpressionsNotInAsyncContext();
      return s;
    }
  }, {
    key: "reduceFunctionBody",
    value: function reduceFunctionBody(node, _ref10) {
      var directives = _ref10.directives,
          statements = _ref10.statements;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceFunctionBody", this).call(this, node, { directives: directives, statements: statements });
      s = s.clearFreeReturnStatements();
      return s;
    }
  }, {
    key: "reduceFunctionDeclaration",
    value: function reduceFunctionDeclaration(node, _ref11) {
      var name = _ref11.name,
          params = _ref11.params,
          body = _ref11.body;

      body = node.isGenerator ? body.clearYields() : body.enforceYields();
      body = node.isAsync ? body.clearAwaitExpressionsNotInAsyncContext() : body.enforceAwaitExpressionsNotInAsyncContext();
      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceFunctionDeclaration", this).call(this, node, { name: name, params: params, body: body });
      if (node.isGenerator && node.isAsync) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.ASYNC_GENERATOR_FUNCTION));
      }
      return s;
    }
  }, {
    key: "reduceFunctionExpression",
    value: function reduceFunctionExpression(node, _ref12) {
      var name = _ref12.name,
          params = _ref12.params,
          body = _ref12.body;

      body = node.isGenerator ? body.clearYields() : body.enforceYields();
      body = node.isAsync ? body.clearAwaitExpressionsNotInAsyncContext() : body.enforceAwaitExpressionsNotInAsyncContext();
      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceFunctionExpression", this).call(this, node, { name: name, params: params, body: body });
      if (node.isGenerator && node.isAsync) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.ASYNC_GENERATOR_FUNCTION));
      }
      return s;
    }
  }, {
    key: "reduceGetter",
    value: function reduceGetter(node, _ref13) {
      var name = _ref13.name,
          body = _ref13.body;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceGetter", this).call(this, node, { name: name, body: body.enforceYields().enforceAwaitExpressionsNotInAsyncContext() });
      return s;
    }
  }, {
    key: "reduceIdentifierExpression",
    value: function reduceIdentifierExpression(node) {
      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceIdentifierExpression", this).call(this, node);
      if (!isValidIdentifier(node.name)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_IDENTIFIER_NAME));
      }
      return s;
    }
  }, {
    key: "reduceIfStatement",
    value: function reduceIfStatement(node, _ref14) {
      var test = _ref14.test,
          consequent = _ref14.consequent,
          alternate = _ref14.alternate;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceIfStatement", this).call(this, node, { test: test, consequent: consequent, alternate: alternate });
      if (isProblematicIfStatement(node)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_IF_STATEMENT));
      }
      if (node.consequent.type === 'ClassDeclaration' || node.consequent.type === 'FunctionDeclaration' && node.consequent.isGenerator || node.consequent.type === 'VariableDeclarationStatement' && (node.consequent.declaration.kind === 'let' || node.consequent.declaration.kind === 'const')) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.PROPER_DECLARATION_AS_STATEMENT));
      }
      if (node.alternate && (node.alternate.type === 'ClassDeclaration' || node.alternate.type === 'FunctionDeclaration' && node.alternate.isGenerator || node.alternate.type === 'VariableDeclarationStatement' && (node.alternate.declaration.kind === 'let' || node.alternate.declaration.kind === 'const'))) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.PROPER_DECLARATION_AS_STATEMENT));
      }
      return s;
    }
  }, {
    key: "reduceImportSpecifier",
    value: function reduceImportSpecifier(node, _ref15) {
      var binding = _ref15.binding;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceImportSpecifier", this).call(this, node, { binding: binding });
      if (node.name !== null && !isIdentifierNameES6(node.name)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_IMPORT_SPECIFIER_NAME));
      }
      return s;
    }
  }, {
    key: "reduceLabeledStatement",
    value: function reduceLabeledStatement(node, _ref16) {
      var body = _ref16.body;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceLabeledStatement", this).call(this, node, { body: body });
      if (!isValidIdentifier(node.label)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_LABEL));
      }
      s = checkIllegalBody(node, s, { allowFunctions: true });
      return s;
    }
  }, {
    key: "reduceLiteralNumericExpression",
    value: function reduceLiteralNumericExpression(node) {
      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceLiteralNumericExpression", this).call(this, node);
      if (isNaN(node.value)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.LITERAL_NUMERIC_VALUE_NOT_NAN));
      }
      if (node.value < 0 || node.value == 0 && 1 / node.value < 0) {
        // second case is for -0
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.LITERAL_NUMERIC_VALUE_NOT_NEGATIVE));
      }
      if (!isNaN(node.value) && !isFinite(node.value)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.LITERAL_NUMERIC_VALUE_NOT_INFINITE));
      }
      return s;
    }
  }, {
    key: "reduceLiteralRegExpExpression",
    value: function reduceLiteralRegExpExpression(node) {
      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceLiteralRegExpExpression", this).call(this, node);
      if (!(0, _shiftRegexpAcceptor2.default)(node.pattern, { unicode: node.unicode })) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_REG_EX_PATTERN));
      }
      return s;
    }
  }, {
    key: "reduceMethod",
    value: function reduceMethod(node, _ref17) {
      var params = _ref17.params,
          body = _ref17.body,
          name = _ref17.name;

      body = node.isGenerator ? body.clearYields() : body.enforceYields();
      body = node.isAsync ? body.clearAwaitExpressionsNotInAsyncContext() : body.enforceAwaitExpressionsNotInAsyncContext();
      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceMethod", this).call(this, node, { params: params, body: body, name: name });
      if (node.isGenerator && node.isAsync) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.ASYNC_GENERATOR_FUNCTION));
      }
      return s;
    }
  }, {
    key: "reduceModule",
    value: function reduceModule(node, _ref18) {
      var directives = _ref18.directives,
          items = _ref18.items;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceModule", this).call(this, node, { directives: directives, items: items });
      s = s.enforceFreeReturnStatements();
      s = s.enforceBindingIdentifiersCalledDefault();
      s = s.enforceYields();
      s = s.enforceAwaitExpressionsNotInAsyncContext();
      return s;
    }
  }, {
    key: "reduceReturnStatement",
    value: function reduceReturnStatement(node, _ref19) {
      var expression = _ref19.expression;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceReturnStatement", this).call(this, node, { expression: expression });
      s = s.addFreeReturnStatement(node);
      return s;
    }
  }, {
    key: "reduceScript",
    value: function reduceScript(node, _ref20) {
      var directives = _ref20.directives,
          statements = _ref20.statements;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceScript", this).call(this, node, { directives: directives, statements: statements });
      s = s.enforceFreeReturnStatements();
      s = s.enforceBindingIdentifiersCalledDefault();
      s = s.enforceYields();
      s = s.enforceAwaitExpressionsNotInAsyncContext();
      return s;
    }
  }, {
    key: "reduceSetter",
    value: function reduceSetter(node, _ref21) {
      var name = _ref21.name,
          param = _ref21.param,
          body = _ref21.body;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceSetter", this).call(this, node, { name: name, param: param, body: body.enforceYields().enforceAwaitExpressionsNotInAsyncContext() });
      return s;
    }
  }, {
    key: "reduceStaticMemberExpression",
    value: function reduceStaticMemberExpression(node, _ref22) {
      var object = _ref22.object;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceStaticMemberExpression", this).call(this, node, { object: object });
      if (!isIdentifierNameES6(node.property)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_STATIC_MEMBER_EXPRESSION_PROPERTY_NAME));
      }
      return s;
    }
  }, {
    key: "reduceStaticMemberAssignmentTarget",
    value: function reduceStaticMemberAssignmentTarget(node, _ref23) {
      var object = _ref23.object;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceStaticMemberAssignmentTarget", this).call(this, node, { object: object });
      if (!isIdentifierNameES6(node.property)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_STATIC_MEMBER_ASSIGNMENT_TARGET_PROPERTY_NAME));
      }
      return s;
    }
  }, {
    key: "reduceTemplateElement",
    value: function reduceTemplateElement(node) {
      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceTemplateElement", this).call(this, node);
      if (!isTemplateElement(node.rawValue)) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.VALID_TEMPLATE_ELEMENT_VALUE));
      }
      return s;
    }
  }, {
    key: "reduceTemplateExpression",
    value: function reduceTemplateExpression(node, _ref24) {
      var tag = _ref24.tag,
          elements = _ref24.elements;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceTemplateExpression", this).call(this, node, { tag: tag, elements: elements });
      if (node.elements.length > 0) {
        if (node.elements.length % 2 == 0) {
          s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.ALTERNATING_TEMPLATE_EXPRESSION_ELEMENTS));
        } else {
          node.elements.forEach(function (x, i) {
            if (i % 2 == 0) {
              if (!(x.type === 'TemplateElement')) {
                s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.ALTERNATING_TEMPLATE_EXPRESSION_ELEMENTS));
              }
            } else {
              if (x.type === 'TemplateElement') {
                s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.ALTERNATING_TEMPLATE_EXPRESSION_ELEMENTS));
              }
            }
          });
        }
      }
      return s;
    }
  }, {
    key: "reduceVariableDeclaration",
    value: function reduceVariableDeclaration(node, _ref25) {
      var declarators = _ref25.declarators;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceVariableDeclaration", this).call(this, node, { declarators: declarators });
      if (node.declarators.length == 0) {
        s = s.addError(new _validationContext.ValidationError(node, _validationErrors2.default.NOT_EMPTY_VARIABLE_DECLARATORS_LIST));
      }
      return s;
    }
  }, {
    key: "reduceWhileStatement",
    value: function reduceWhileStatement(node, _ref26) {
      var test = _ref26.test,
          body = _ref26.body;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceWhileStatement", this).call(this, node, { test: test, body: body });
      s = checkIllegalBody(node, s);
      return s;
    }
  }, {
    key: "reduceWithStatement",
    value: function reduceWithStatement(node, _ref27) {
      var object = _ref27.object,
          body = _ref27.body;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceWithStatement", this).call(this, node, { object: object, body: body });
      s = checkIllegalBody(node, s);
      return s;
    }
  }, {
    key: "reduceYieldExpression",
    value: function reduceYieldExpression(node, _ref28) {
      var expression = _ref28.expression;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceYieldExpression", this).call(this, node, { expression: expression });
      s = s.addYieldExpressionNotInGeneratorContext(node);
      return s;
    }
  }, {
    key: "reduceAwaitExpression",
    value: function reduceAwaitExpression(node, _ref29) {
      var expression = _ref29.expression;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceAwaitExpression", this).call(this, node, { expression: expression });
      s = s.addAwaitExpressionNotInAsyncContext(node);
      return s;
    }
  }, {
    key: "reduceYieldGeneratorExpression",
    value: function reduceYieldGeneratorExpression(node, _ref30) {
      var expression = _ref30.expression;

      var s = _get(Validator.prototype.__proto__ || Object.getPrototypeOf(Validator.prototype), "reduceYieldGeneratorExpression", this).call(this, node, { expression: expression });
      s = s.addYieldGeneratorExpressionNotInGeneratorContext(node);
      return s;
    }
  }], [{
    key: "validate",
    value: function validate(node) {
      return (0, _shiftReducer2.default)(new Validator(), node).errors.concat(_shiftParser.EarlyErrorChecker.check(node));
    }
  }]);

  return Validator;
}(_shiftReducer.MonoidalReducer);
