// deobf/dist/modifications/execution/scope.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Scope {
    /**
     * Creates a new scope.
     * @param node The node that created the scope.
     * @param parent The parent scope (optional).
     */
    constructor(node, parent) {
        this.node = node;
        this.parent = parent;
        this.children = new Map();
        this.executedFunctions = new Map();
        if (this.parent) {
            this.parent.children.set(this.node, this);
        }
    }
    /**
     * Searches for an executed function by name.
     * @param name The name of the executed function.
     */
    findExecutedFunction(name) {
        if (this.executedFunctions.has(name)) {
            return this.executedFunctions.get(name);
        }
        return this.parent ? this.parent.findExecutedFunction(name) : null;
    }
    /**
     * Adds an executed function.
     * @param array The executed function to be added.
     */
    addExecutedFunction(func) {
        if (func.name) {
            this.executedFunctions.set(func.name, func);
        }
    }
    /**
     * Adds an alias for an executed function.
     * @param func The executed function.
     * @param name The alias.
     */
    addAlias(func, name) {
        this.executedFunctions.set(name, func);
    }
}
exports.default = Scope;
