// @shineiichijo/marika/dist/constants/anime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnimeStatus = exports.Ratings = exports.AnimeTypes = void 0;
var AnimeTypes;
(function (AnimeTypes) {
    AnimeTypes["TV"] = "TV";
    AnimeTypes["MOVIE"] = "Movie";
    AnimeTypes["SPECIAL"] = "Special";
    AnimeTypes["OVA"] = "OVA";
    AnimeTypes["ONA"] = "ONA";
    AnimeTypes["MUSIC"] = "Music";
})(AnimeTypes || (exports.AnimeTypes = AnimeTypes = {}));
var Ratings;
(function (Ratings) {
    /** G - All Ages */
    Ratings["G"] = "g";
    /** PG - Children */
    Ratings["PG"] = "pg";
    /** PG-13 - Teens 13 or older */
    Ratings["PG_13"] = "pg13";
    /** R - 17+ (violence & profanity) */
    Ratings["R_17"] = "r17";
    /** R+ - Mild Nudity */
    Ratings["R"] = "r";
    /** Rx - Hentai */
    Ratings["RX"] = "rx";
})(Ratings || (exports.Ratings = Ratings = {}));
var AnimeStatus;
(function (AnimeStatus) {
    AnimeStatus["AIRING"] = "airing";
    AnimeStatus["COMPLETE"] = "complete";
    AnimeStatus["UPCOMING"] = "upcoming";
})(AnimeStatus || (exports.AnimeStatus = AnimeStatus = {}));
