// @shineiichijo/marika/dist/constants/clubs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClubCategories = exports.ClubAccess = void 0;
var ClubAccess;
(function (ClubAccess) {
    ClubAccess["PUBLIC"] = "public";
    ClubAccess["PRIVATE"] = "private";
    ClubAccess["SECRET"] = "secret";
})(ClubAccess || (exports.ClubAccess = ClubAccess = {}));
var ClubCategories;
(function (ClubCategories) {
    ClubCategories["ANIME"] = "anime";
    ClubCategories["MANGA"] = "manga";
    ClubCategories["ACTORS_AND_ARTISTS"] = "actors_and_artists";
    ClubCategories["CHARACTERS"] = "characters";
    ClubCategories["CITIES_AND_NEIGHBORHOODS"] = "cities_and_neighborhoods";
    ClubCategories["COMPANIES"] = "companies";
    ClubCategories["CONVENTIONS"] = "conventions";
    ClubCategories["GAMES"] = "games";
    ClubCategories["JAPAN"] = "japan";
    ClubCategories["MUSIC"] = "music";
    ClubCategories["OTHER"] = "other";
    ClubCategories["SCHOOLS"] = "schools";
})(ClubCategories || (exports.ClubCategories = ClubCategories = {}));
