// @shineiichijo/marika/dist/constants/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Forums = exports.Sorting = exports.API_URL = void 0;
exports.API_URL = 'https://api.jikan.moe/v4';
var Sorting;
(function (Sorting) {
    Sorting["DESCENDING"] = "desc";
    Sorting["ASCENDING"] = "asc";
})(Sorting || (exports.Sorting = Sorting = {}));
var Forums;
(function (Forums) {
    Forums["ALL"] = "all";
    Forums["EPISODE"] = "episode";
    Forums["OTHER"] = "other";
})(Forums || (exports.Forums = Forums = {}));
__exportStar(require("./anime"), exports);
__exportStar(require("./clubs"), exports);
__exportStar(require("./genres"), exports);
__exportStar(require("./manga"), exports);
__exportStar(require("./schedules"), exports);
__exportStar(require("./seasons"), exports);
__exportStar(require("./users"), exports);
__exportStar(require("./top"), exports);
